rust/arrow/src/array/array.rs

```diff
@@ -1647,6 +1671,221 @@ impl From<(Vec<(Field, ArrayRef)>, Buffer, usize)> for StructArray {
+    ptr: *const T,
+    i: usize,
+    len: usize,
+}
+
+impl<'a, T> std::iter::Iterator for NullableIter<'a, T>
+where
+    T: Clone,
+{
+    type Item = Option<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let i = self.i;
+        if i >= self.len {
+            None
+        } else if self.data.is_null(i) {
+            self.i += 1;
+            Some(None)
+        } else {
+            self.i += 1;
+            unsafe { Some(Some((&*self.ptr.offset(i as isize)).clone())) }
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.len, Some(self.len))
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        let i = self.i;
+        if i + n >= self.len {
+            self.i = self.len;
+            None
+        } else if self.data.is_null(i + n) {
+            self.i += n + 1;
+            Some(None)
+        } else {
+            self.i += n + 1;
+            unsafe { Some(Some((&*self.ptr.offset((i + n) as isize)).clone())) }
+        }
+    }
+}
+
+impl<'a, K: ArrowPrimitiveType> DictionaryArray<K> {
+    /// Return an iterator to the keys of this dictionary.
+    pub fn keys(&'a self) -> NullableIter<'a, K::Native> {
+        NullableIter::<'a, K::Native> {
+            data: &self.data,
+            ptr: unsafe { self.raw_values.get().offset(self.data.offset() as isize) },
+            i: 0,
+            len: self.data.len(),
+        }
+    }
+
+    /// Returns an `ArrayRef` to the dictionary values.
+    pub fn values(&self) -> ArrayRef {
+        self.values.clone()
+    }
+
+    /// Returns a clone of the value type of this list.
+    pub fn value_type(&self) -> DataType {
+        self.values.data().data_type().clone()
+    }
+
+    /// The length of the dictionary is the length of the keys array.
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn is_ordered(&self) -> bool {
+        self.is_ordered
+    }
+}
+
+/// Constructs a `DictionaryArray` from an array data reference.
+impl<T: ArrowPrimitiveType> From<ArrayDataRef> for DictionaryArray<T> {
+    fn from(data: ArrayDataRef) -> Self {
+        assert_eq!(
+            data.buffers().len(),
+            1,
+            "DictionaryArray data should contain a single buffer only (keys)."
+        );
+        assert_eq!(
+            data.child_data().len(),
+            1,
+            "DictionaryArray should contain a single child array (values)."
+        );
+
+        let raw_values = data.buffers()[0].raw_data();
+        let dtype: &DataType = data.data_type();
+        let values = make_array(data.child_data()[0].clone());
+        if let DataType::Dictionary(_, _) = dtype {
+            Self {
+                data: data,
+                raw_values: RawPtrBox::new(raw_values as *const T::Native),
+                values: values,
+                is_ordered: false,
+            }
+        } else {
+            panic!("DictionaryArray must have Dictionary data type.")
```

Comment:
Do you think we should be implementing `TryFrom` instead here (not this PR just asking)?

Reply:
We need to have a review of error handling generally. Far too many panics about. Maybe raise another issue for this.

Reply:
Yep, agreed.